File: DotNet/Kyu6/RectangleIntoSquares.cs
// https://www.codewars.com/kata/rectangle-into-squares/train/csharp

namespace DotNet.Kyu6;

using System.Collections.Generic;
using System;
using System.Linq;

public class SqInRect {
    
    public static List<int> sqInRect(int lng, int wdth) {
        if (lng == wdth)
        {
            return null;
        }

        List<int> result = [];
        while (lng != wdth)
        {
            int minSide = Math.Min(lng, wdth);
            int diffSide = Math.Abs(lng - wdth);
            lng = minSide;
            wdth = diffSide;
            result.Add(minSide);
        }
        result.Add(result.Last());

        return result;
    }
}

public class RectangleIntoSquares
{
    public static void Main()
    {
        Console.WriteLine(string.Join(", ", SqInRect.sqInRect(5, 3)));
        Console.WriteLine(string.Join(", ", SqInRect.sqInRect(20, 14)));
        Console.WriteLine(SqInRect.sqInRect(5, 5));
    }
}

File: Elixir/Kyu6/rectangle-into-squares.ex
# https://www.codewars.com/kata/rectangle-into-squares/train/elixir

defmodule Rec2sq do
  defp _squares_in_rect(side, side), do: [side]

  defp _squares_in_rect(length, width) do
    min_side = min(width, length)
    diff_side = abs(length - width)
    [min_side] ++ _squares_in_rect(min_side, diff_side)
  end

  def squares_in_rect(side, side), do: nil

  def squares_in_rect(length, width), do: _squares_in_rect(length, width)
end

IO.inspect(Rec2sq.squares_in_rect(5, 5))
IO.inspect(Rec2sq.squares_in_rect(5, 3))
IO.inspect(Rec2sq.squares_in_rect(3, 5))
